apidef/oas/event.go

```diff
@@ -0,0 +1,198 @@
+	Trigger event.Event `json:"trigger" bson:"trigger"`
+	// Kind specifies the action to be taken on the event trigger.
+	Kind Kind `json:"type" bson:"type"` // json tag is changed as per contract
+	// ID is the ID of event handler in storage.
+	ID string `json:"id,omitempty" bson:"id,omitempty"`
+	// Name is the name of event handler
+	Name string `json:"name,omitempty" bson:"name,omitempty"`
+
+	Webhook WebhookEvent `bson:"-" json:"-"`
+}
+
+// MarshalJSON marshals EventHandler as per Tyk OAS API definition contract.
+func (e *EventHandler) MarshalJSON() ([]byte, error) {
+	outMap, err := reflect.Cast[map[string]interface{}](*e)
+	if err != nil {
+		return nil, err
+	}
+
+	webhookMap, err := reflect.Cast[map[string]interface{}](e.Webhook)
+	if err != nil {
+		return nil, err
+	}
+
+	outMapVal := *outMap
+	for k, v := range *webhookMap {
+		outMapVal[k] = v
+	}
+
+	return json.Marshal(outMapVal)
+}
+
+// UnmarshalJSON unmarshal EventHandler as per Tyk OAS API definition contract.
+func (e *EventHandler) UnmarshalJSON(in []byte) error {
+	type helperEvent EventHandler
+	helper := helperEvent{}
+	if err := json.Unmarshal(in, &helper); err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(in, &helper.Webhook); err != nil {
+		return err
+	}
+
+	*e = EventHandler(helper)
+	return nil
+}
+
+// WebhookEvent stores the core information about a webhook event.
+type WebhookEvent struct {
+	// URL is the target URL for the webhook.
+	URL string `json:"url" bson:"url"`
+	// Method is the HTTP method for the webhook.
+	Method string `json:"method" bson:"method"`
+	// CoolDownPeriod is the cool-down for the event, so it does not trigger again (in seconds).
+	CoolDownPeriod int64 `json:"coolDownPeriod" bson:"coolDownPeriod"`
+	// BodyTemplate is the template to be used for request payload.
+	BodyTemplate string `json:"bodyTemplate,omitempty" bson:"bodyTemplate,omitempty"`
+	// Headers are the list of request headers to be used.
+	Headers Headers `json:"headers,omitempty" bson:"headers,omitempty"`
+}
+
+// GetWebhookConf converts EventHandler.WebhookEvent apidef.WebHookHandlerConf.
+func (e *EventHandler) GetWebhookConf() apidef.WebHookHandlerConf {
+	return apidef.WebHookHandlerConf{
+		Disabled:     !e.Enabled,
+		ID:           e.ID,
+		Name:         e.Name,
+		Method:       e.Webhook.Method,
+		TargetPath:   e.Webhook.URL,
+		HeaderList:   e.Webhook.Headers.Map(),
+		EventTimeout: e.Webhook.CoolDownPeriod,
+		TemplatePath: e.Webhook.BodyTemplate,
+	}
+}
+
+// EventHandlers holds the list of events to be processed for the API.
+type EventHandlers []EventHandler
+
+// Fill fills EventHandlers from classic API definition. Currently only webhook events are supported.
+func (e *EventHandlers) Fill(api apidef.APIDefinition) {
+	if len(api.EventHandlers.Events) == 0 {
+		return
+	}
+
+	events := EventHandlers{}
+	for gwEvent, ehs := range api.EventHandlers.Events {
+		for _, eh := range ehs {
+			if eh.Handler != event.WebHookHandler {
```

Comment:
also should be a switch + extract the nested code for webhook handler into internal `fillWebhook` so the cognitive complexity goes down, and we're covered with linters